onnxruntime/core/providers/cuda/cuda_execution_provider.cc

```diff
@@ -359,6 +363,25 @@ Status CUDAExecutionProvider::SetComputeStream(void* stream) {
   return Status::OK();
 }
 
+#if defined(CUDA_VERSION) && CUDA_VERSION >= 11000
+void CUDAExecutionProvider::CaptureBegin()  {
+  ORT_THROW_IF_ERROR(Sync());
+  graph_.CaptureBegin();
+}
+
+void CUDAExecutionProvider::CaptureEnd() {
+  graph_.CaptureEnd();
+  ORT_THROW_IF_ERROR(Sync());
+}
+
+void CUDAExecutionProvider::Replay() {
+  ORT_THROW_IF_ERROR(OnRunStart());
```

Comment:
Curious - Do we need OnRunStart() and OnRunEnd() while replaying the graph or can we just move the stream syncing in Graph::Replay() and just call Graph::Replay() ?

Reply:
Yes, we can move the stream syncing into Graph::Replay().

Reply:
Actually syncing is not needed inside replay if the input tensors have been updated.

Reply:
But a stream sync is needed OnRunEnd() so that we return the control back to the caller only after the replay is done right ?

Reply:
You are correct. `cudaStreamSynchronize` is added inside `Status CUDAGraph::Replay()`